### DISCUSSION TITLE
Would it be viable to un-nullify Nullable<T> values once they are guaranteed to be non-null?

### DISCUSSION THREAD

**Discussion by Gnbrkm41** [dotnet/csharplang#7234] [Language Ideas] [answered]
Say we have this code, which takes a nullable reference type object and returns a non-nullable reference type object:
```cs
string Foo(string? a)
{
    if (a == null)
    {
        a = FunctionThatReturnsNonNull();
    }
    return a;
}
```
Because `a` is known to be non-null in compile time, `a` is treated as a non-nullable `string` after the if block, and will compile without warnings.

However, the same is not the case with nullable value types:
```cs
int Foo(int? a)
{
    if (a == null)
    {
        a = FunctionThatReturnsNonNull();
    }
    return a;
}
```

This is because even though we know `a` is probably not null at compile time after the if block, unlike NRTs nullable value types are actually encased in a `Nullable<T>` and therefore is not implicitly convertible to an `int`. Users would need to insert an explicit `.Value` call or an `(int)` cast to make the code compile.

Because NRT is an afterthought and `Nullable<T>` precedes NRT by a fair bit, I understand why this is the case. However it would be nice if the `Nullable<T>` values can be implicitly converted to the underlying `T` value (i.e. have its `.Value` property inserted automatically) once the compiler can infer its nullability from usage?

The biggest problem would be overload resolutions, though. If there is a method taking both `int` and `object` (or any other interfaces it can box to for that matter) for example, currently it would select the `Function(object)` instead of the (probably) desired `int`. If we allow implicit conversions this would make above code call into a different method... I imagine there isn't really a good way to get around this, and this makes me very sad.

It's a shame that we didn't do this earlier, because it really just feels "wrong" to having to type `.Value` even though it is guaranteed to be non-null! (evil joke thoughts - how about making `!`s after a `Nullable<T>` call `.Value` 😄)

**Comment by 333fred** [comment, marked as answer]
We talked about this during the development of the nrt feature: https://github.com/dotnet/csharplang/issues/1865#issuecomment-544685799. And you're correct that the challenges around reinterpretation of existing code were the blockers.

**Reply by TahirAhmadov** [reply]
Yes, from a clean slate it would have been much preferable, but just plain dangerous at this point, unfortunately. It's nice that the team considered it seriously, though.
